### DISCUSSION TITLE
Using Geoserver to display active data in a shapefile feature

### DISCUSSION THREAD

**Discussion by mlandsfeld** [tethysplatform/tethys#652] [Q&A] [answered]
I am developing a Tethys app that displays crop yield estimates by administrative districts in East Africa. I am using a shapefile that contains the estimates for each district for the last 20 years. I have loaded the shapefile into a GeoServer and have made a few test layers by creating an SLD for each of the years where the ogc:PropertyName is the year, which is also 
the name of the column in the database DBF file. I have combined the styles with the data store to create a layer of each year. This will be a continually updated viewer and there will be other variables as well. I don't want to have to create layers by hand every month as new estimates come in. I am wondering if there is a way to style the store on the
fly with the new estimates or a way to automate the layer creation. This may be the wrong approach though if there is a way to programatically use the values from a database and  inject those into the GeoServer to as the attribute values for the feature store. 
Any ideas or pointers appreciated.

**Comment by swainn** [comment, marked as answer]
@mlandsfeld it is definitely possible to modify/manage GeoServer on the fly/programmatically. GeoServer has a [REST API](https://docs.geoserver.org/stable/en/user/rest/) that can be used to manage all the data in GeoServer including workspaces, stores, layers, and styles. The official way to interact with GeoServer in Tethys is through the SpatialDatasetServices API. Part of the API is the GeoServer SpatialDatasetEngine, which provides a Python wrapper around the GeoServer REST API. Here are links to the pertinent documentation:

* [SpatialDatasetServices API](http://docs.tethysplatform.org/en/stable/tethys_sdk/tethys_services/spatial_dataset_services.html)
* [GeoServer SpatialDatasetEngine Reference](http://docs.tethysplatform.org/en/stable/tethys_sdk/tethys_services/spatial_dataset_service/geoserver_reference.html)

To add new Shapefiles to the GeoServer using GeoServer SpatialDatasetEngine you can use the `create_shapefile_resource()` method as illustrated in this example: http://docs.tethysplatform.org/en/stable/tethys_sdk/tethys_services/spatial_dataset_service/geoserver_reference.html#upload-shapefile

For Styles, I'd recommend you look at creating a dynamic style that leverages GeoServer's variable substitution, which will allow you to do things like change the colors or breakpoints through the URL parameters of the layer request. See: [Variable Substitution](https://docs.geoserver.org/stable/en/user/styling/sld/extensions/substitution.html).

NOTE: The Spatial Dataset Services API has been designed to work with the version of GeoServer provided by the Docker image you can install with Tethys (I believe it is version 2.08 at the time of writing). It probably won't work on newer versions of GeoServer. However, there are other Python packages for working with the GeoServer REST API that would work equally well:

* [gsconfig-py3](https://pypi.org/project/gsconfig-py3/)
* [geoserver-rest](https://pypi.org/project/geoserver-rest/)

Of course, you can also use the [GeoServer REST API](https://docs.geoserver.org/stable/en/user/rest/) directly using something like the Python [requests](https://pypi.org/project/requests/) module.
